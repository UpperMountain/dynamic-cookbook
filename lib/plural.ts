function round(amt: number, accuracy: number): number {
  // Round to closest multiple of `round`.
  return Math.round(amt / accuracy) * accuracy;
}

function format(amt: number): string {
  // Don't display decimal places for integers. Otherwise, display two.
  const formatted: string =
    amt % 1 === 0
      ? amt.toString()
      : Math.abs((amt * 100) % 10) < 1 // valid for 0.500111
      ? amt.toFixed(1)
      : amt.toFixed(2);

  return formatted;
}

export function qty(
  amt: number,
  accuracy: number,
  singular?: string,
  plural?: string
) {
  const rounded = round(amt, accuracy);
  const formatted = format(rounded);

  if (!singular && !plural) {
    return formatted;
  }

  if (rounded === 1 || !plural) {
    return `${formatted} ${singular}`;
  } else {
    return `${formatted} ${plural}`;
  }
}

export function plural(
  amt: number,
  accuracy: number,
  singular: string,
  plural: string
) {
  const rounded = round(amt, accuracy);
  if (rounded === 1) {
    return singular;
  } else {
    return plural;
  }
}
