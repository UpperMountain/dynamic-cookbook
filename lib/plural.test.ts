import { qty, plural } from "./plural";

describe("qty()", () => {
  it("works without exploding", () => {
    qty(14, 1, "apple", "apples");
  });

  it("passes snapshot tests", () => {
    expect([
      qty(12, 1, "potato", "potatoes"),
      qty(1, 1, "potato", "potatoes"),
      qty(1.234422222, 1, "tsp"),
      qty(124.3443, 0.5),
      qty(12.5, 1, "potato", "potatoes"),
      qty(-12.5, 1, "potato", "potatoes"),
      qty(-12.344343434, 0.25, "potato", "potatoes"),
      qty(-12.344343434, 0.5, "potato", "potatoes")
    ]).toMatchSnapshot();
  });
});

describe("plural()", () => {
  it("works without exploding", () => {
    plural(14, 1, "apple", "apples");
  });

  it("passes snapshot tests", () => {
    expect([
      plural(12, 1, "potato", "potatoes"),
      plural(1, 1, "potato", "potatoes"),
      plural(12.5, 1, "potato", "potatoes"),
      plural(-12.5, 1, "potato", "potatoes"),
      plural(1.2, 1, "potato", "potatoes"),
      plural(1, 0.5, "potato", "potatoes")
    ]).toMatchSnapshot();
  });
});
