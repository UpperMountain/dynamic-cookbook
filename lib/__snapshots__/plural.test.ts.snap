// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`plural() passes snapshot tests 1`] = `
Array [
  "potatoes",
  "potato",
  "potatoes",
  "potatoes",
  "potato",
  "potato",
]
`;

exports[`qty() passes snapshot tests 1`] = `
Array [
  "12 potatoes",
  "1 potato",
  "1 tsp",
  "124.5",
  "13 potatoes",
  "-12 potatoes",
  "-12.25 potatoes",
  "-12.5 potatoes",
]
`;
